pipelinerl/async_llm.py

```diff
@@ -72,26 +111,111 @@ async def llm_async_generate(llm: TrainableLLM, prompt: Prompt, session: aiohttp
 
     output = LLMOutput(content=content)
     llm_call = llm.log_output(prompt, output, count_tokens=False)
-    llm_call.prompt_length_tokens = data['usage']['prompt_tokens']
-    llm_call.output_length_tokens = data['usage']['completion_tokens']
+    llm_call.prompt_length_tokens = data["usage"]["prompt_tokens"]
+    llm_call.output_length_tokens = data["usage"]["completion_tokens"]
     assert llm_call is not None, "llm_call is None"
     llm_call.logprobs = parsed_logprobs
     return llm_call
 
 
 def make_training_text(llm: TrainableLLM, llm_call: LLMCall) -> TrainingText:
-    training_text = llm.make_training_text(llm_call.prompt, llm_call.output)
+    # Extract visual features if present
+    pixel_values = None
+    image_thw = None
+    images = []
+    use_processor = False
+    full_messages = llm_call.prompt.messages + [{"role": "assistant", "content": llm_call.output.content}]
+    
+    if hasattr(llm_call.prompt, "messages"):
+        images = extract_images_from_messages(llm_call.prompt.messages)
+        if images:
+            use_processor = True
+    
+    if use_processor:
+        # Use processor for vision-language models
+        processor = processor_factory.get_processor(llm.model_name)
+        
+        try:
+            # Apply chat template using processor for proper image token handling
+            prompt_text = processor.apply_chat_template(
+                llm_call.prompt.messages,
+                tokenize=False,
+                add_generation_prompt=True,
+            )
+            
+            # Create full conversation with assistant response
+            text = processor.apply_chat_template(
+                full_messages,
+                tokenize=False,
+            )
+            
+            # Process prompt with images to get token IDs with image placeholders
+            prompt_inputs = processor(
+                text=processor.apply_chat_template(llm_call.prompt.messages, tokenize=False, add_generation_prompt=True),
+                images=images,
+                return_tensors=None,
+            )
+            
+            # prompt_inputs["input_ids"] is a list of list
+            prompt_token_ids = prompt_inputs["input_ids"][0]
+            
+            # Process images to get visual features
+            processed = processor(
+                text=[prompt_text], 
+                images=images, 
+                padding=True, 
+                return_tensors=None
+            )
+            pixel_values = processed.pixel_values.astype(np.float16)
```

Comment:
I'd double check that `np.float32` and `np.float16` give the same downstream perfromance.